content/guides/typescript-for-solid.mdx

```diff
@@ -0,0 +1,714 @@
+
+Consider a situation where you want to show a name only when it's defined using a Solid signal:
+
+```typescript
+const [name, setName] = createSignal<string>();
+return (
+  <Show when={(() => {
+      const v = data()
+      return v && v.type === "user" && v // T: User | undefined
+    })()}>
+  )
+```
+
+Accessing the callback may throw an error if the `when` prop is nullish or false at the time of access.
+Usually this is not the case, however in `onCleanup` or asynchronous contexts it can be.
+
+TypeScript may not like this.
+While you can understand that the second call to `name()` happens only if the first is truthy, TypeScript cannot. 
+It may flag an error indicating that `name()` could potentially be `undefined` when `.replace()` is called.
+
+There are a few ways to tackle this:
+
+1. **Non-Null Assertion**
+Use TypeScript's [non-null assertion operator (!).](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#non-null-assertion-operator-postfix-) 
+This operator tells TypeScript you're confident the value isn't `null` or `undefined`:
+
+```typescript
+return <Show when={name()}>Hello {name()!.replace(/\s+/g, "\xa0")}!</Show>;
+```
+While this may be a quick option, incorrectly asserting can lead to runtime errors if the value does turn out to be `null` or `undefined`.
+
+2. **Callback form of Show**
+
+Solid provides a callback form for its [`<Show>`](/reference/show) component. 
+In this form, the value that makes the when prop true is passed into the component:
+
+```typescript
+const [data, setData] = createSignal<User | Admin>() // T: Signal<User | Admin | undefined>
+
+return(
+    <Show when={(() => {
+      const v = data()
+      return v && v.type === "user" && v // T: User | undefined
+    })()}>
+  )
+```
+
+This approach takes advantage of the type information embedded in the [`<Show>`](/reference/show) component. 
+By temporarily storing the value `v`, you can use type narrowing to specify its type more accurately before returning it.
+
+You do have to be cautious, however, since accessing the callback may throw an error if the when prop is nullish or false at the time of access. 
+This is usually not an issue, but in asynchronous contexts or during [`onCleanup`](/reference/oncleanup), it may occur.
+
+<Aside>
+If you find the above method a bit shorthand, there's a more explicit alternative. 
+Solid's [`<Show>`](/reference/show) component actually performs memoization, so you don't necessarily need a memo for this:
+
+```tsx
+const user = createMemo(() => {
+  const v = data()
+  return v && v.type === "user" && v // Type: User | undefined
+})
+
+<Show when={user()}>
+  {user => <User user={user()} />}
+</Show>
+```
+While this approach is more explicit, it's worth noting that the primary example for the callback form should ideally not use this technique for the sake of clarity
+</Aside>
```

Comment:
I think this section needs a rewrite. I'll do what I can if I have the time.

Reply:
@otonashixav what makes it need a rewrite?

Reply:
It does a poor job of contextualizing exactly when, and what challenges would be encountered. In other words, it needs to
- Provide an example of why control flow might be problematic in the first place -- usually this would be because a function or component wants a specific type that you want to be able to guarantee, but cannot do so because your values are wrapped in accessors, and won't be narrowed. 
- Show other solutions:
  - Solid wants you to avoid conditional rendering, and prefers if you push it to the leaves, e.g. rewrite your components and functions to accept `undefined` and use either nullish coalescing (`signal() ?? fallback`) or optional chaining (`signal()?.value`).
  - If that isn't possible, or you don't think it's worth the effort, then either `Show` (or `Switch` and `Match`), or other forms of control flow can be used. There are many ways to accomplish this but using the control flow components "properly" will avoid unnecessarily recreating components where possible. 
- Show how `Show` can be used to achieve some semblance of type narrowing in the "official" capacity, i.e. via excluding nullish types, regardless of whether `keyed` is used or not -- `keyed` itself also warrants some explanation since the unkeyed callback form is somewhat restricted in usefulness to typescript.
- Show how the above is insufficient for anything other than removing nullish values -- **the section currently seems to focus solely on this, which is why I believe it should be rewritten.** The `when=IIFE` workaround is not really something I would recommend in an official capacity, so it should be in an aside; I think you misinterpreted my previous comment here. Using a computed signal or memo should be preferred for clarity. The difference between the two can be elaborated elsewhere.

Reply:
Alright. Rewrite the section when you can. I'm not sure I understaand exactly what you're looking for so I don't feel I can convey what you want.